### PR TITLE
SEAB-5387: get notebook counts in collections and add Service support in collections

### PR DESCRIPTION
**Pull request by hyunnaye**
**Description**
This PR adds `notebooksLength` to collections to retrieve the notebooks count for each collection. Also, this PR adds implementation for Services to be added in collections/categories

**Review Instructions**
Review that the tests pass for both notebook counts and services.

**Issue**
https://ucsc-cgl.atlassian.net/browse/SEAB-5387

**Security**
If there are any concerns that require extra attention from the security team, highlight them here.

Please make sure that you've checked the following before submitting your pull request. Thanks!

- [x] Check that you pass the basic style checks and unit tests by running `mvn clean install`
- [x] Ensure that the PR targets the correct branch. Check the milestone or fix version of the ticket.
- [x] Follow the existing JPA patterns for queries, using named parameters, to avoid SQL injection
- [x] If you are changing dependencies, check the Snyk status check or the dashboard to ensure you are not introducing new high/critical vulnerabilities
- [x] Assume that inputs to the API can be malicious, and sanitize and/or check for Denial of Service type values, e.g., massive sizes
- [x] Do not serve user-uploaded binary images through the Dockstore API
- [x] Ensure that endpoints that only allow privileged access enforce that with the `@RolesAllowed` annotation
- [x] Do not create cookies, although this may change in the future
- [x] If this PR is for a user-facing feature, create and link a documentation ticket for this feature (usually in the same milestone as the linked issue). Style points if you create a documentation PR directly and link that instead.

### CODE REVIEW

File: dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/EntryDAO.java
Comment (coverbeck) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/EntryDAO.java:227]: This means notebooks are going to be a subset of workflows, which seems wrong. It seems like they should be distinct.

File: dockstore-webservice/src/main/java/io/dockstore/webservice/resources/CollectionHelper.java
Comment (coverbeck) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/resources/CollectionHelper.java:117]: Here the workflows length will include notebooks, because `getCollectionWorkflows` returns notebooks and workflows. So if you have 2 notebooks and 2 bio workflows, the response will include:

* Workflows length of 4
* Notebooks length of 2

But I would expect

* Workflows length of 2
* Notebooks length of 2

**Review by coverbeck — COMMENTED** [review]
This code is (already was) a little confusing, to me:

* Entry.getCollectionsWorkflows -- returns bioworkflows, apptools, and notebooks
* Entry.getCollectionsTools - returns legacy tools
* Entry.getWorkflowsLength -- returns length of bioworkflows only
* Entry.getToolsLength - returns length of legacy tools

App tools' length is not included in the lengths for the [summary](https://github.com/dockstore/dockstore/blob/9c91f008e0fabc834e9591411662b00ad74ebbf9/dockstore-webservice/src/main/java/io/dockstore/webservice/resources/CollectionHelper.java#L63), as far as I can tell. And are part of the workflow count for the detailed response.

Sorry to bring all this up. Some of this could be split into a new ticket. For now, I think you should make sure that notebooks are not double-counted in the totals.

File: dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/EntryDAO.java
Comment (svonworl) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/EntryDAO.java:227]: A core question here is: "What is a workflow?"  Alas, in our naming scheme, it depends...
In the class hierarchy, `Workflow` is the parent class of `BioWorkflow`, `Notebook`, `AppTool`, and `Service`, and so it's not a concrete workflow in the `BioWorkflow` sense, but an abstract thing, more like an `Entry`-plus.
This produces all sorts of confusion.  We use the term `workflow` to mean "BioWorkflow" in some places, and "Entry-plus" in others. 
IMHO, there should probably be methods to retrieve the collection contents for each of the concrete entry types, and maybe a method that retrieves all entries in the collection.

File: dockstore-webservice/src/main/java/io/dockstore/webservice/core/Collection.java
Comment (denis-yuen) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/core/Collection.java:122]: Two position 6

File: dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/EntryDAO.java
Comment (denis-yuen) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/EntryDAO.java:227]: > This means notebooks are going to be a subset of workflows, which seems wrong. It seems like they should be distinct.

Is it though? Isn't the query `from Workflow w, Collection col join col.entries as e where type(w) in (BioWorkflow, AppTool, Notebook )` i.e.

Comment (denis-yuen) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/EntryDAO.java:227]: > IMHO, there should probably be methods to retrieve the collection contents for each of the concrete entry types, and maybe a method that retrieves all entries in the collection.

Fair

File: dockstore-webservice/src/main/java/io/dockstore/webservice/resources/CollectionHelper.java
Comment (denis-yuen) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/resources/CollectionHelper.java:117]: Could use some tests as well before we go refactoring taking into account @svonworl 's suggestion

File: dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/EntryDAO.java
Comment (coverbeck) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/EntryDAO.java:227]: > > This means notebooks are going to be a subset of workflows, which seems wrong. It seems like they should be distinct.
> 
> Is it though? Isn't the query `from Workflow w, Collection col join col.entries as e where type(w) in (BioWorkflow, AppTool, Notebook )` i.e.

Looks like you may not have finished your thought, but I meant `getCollectionWorkflows()` returns bioworkflows, app tools, and notebooks, while `getCollectionNotebooks()` returns just notebooks. And then further below we set the lengths based on the length of those two collections, which means notebooks would get double counted.

Steve's suggestion of `getCollectionBioWorkflows()`,  etc. would address this.

Comment (hyunnaye) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/EntryDAO.java]: Added methods for each entryTypes. Thank you for everyone's inputs!

File: dockstore-webservice/src/main/java/io/dockstore/webservice/core/Entry.java
Comment (denis-yuen) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/core/Entry.java:94]: Are these three queries the same as `Entry.getAllCollectionWorkflows` but just limited to one of workflow, apptool. or notebook? 

Don't need to do this now, but I feel that there's gotta be a way to use subqueries with a parameter. https://stackoverflow.com/questions/3629259/how-do-i-do-a-jpql-subquery 
Should be new ticket since this PR has been alive for a good while.

File: dockstore-webservice/src/main/java/io/dockstore/webservice/core/Entry.java
Comment (denis-yuen) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/core/Entry.java:105]: Ditto with these

**Review by denis-yuen — APPROVED** [review]
See comment, feel like there's probably a way to simplify. 
Can probably assign that to a follow-up

File: dockstore-webservice/src/main/java/io/dockstore/webservice/core/Entry.java
Comment (hyunnaye) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/core/Entry.java:94]: Ticket: https://ucsc-cgl.atlassian.net/browse/SEAB-5560

File: dockstore-webservice/src/main/java/io/dockstore/webservice/core/Entry.java
Comment (kathy-t) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/core/Entry.java:93]: Missing `when` for services

File: dockstore-webservice/src/main/java/io/dockstore/webservice/core/Entry.java
Comment (kathy-t) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/core/Entry.java:104]: Missing `when` for services

File: dockstore-webservice/src/main/java/io/dockstore/webservice/resources/CollectionHelper.java
Comment (kathy-t) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/resources/CollectionHelper.java:85]: For the switch statement below, it's missing the `"services"` case

File: dockstore-webservice/src/main/java/io/dockstore/webservice/resources/CollectionHelper.java
Comment (kathy-t) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/resources/CollectionHelper.java:76]: This gets all app tools, bioworkflows, notebooks and services. On line 113 - 118, you also get the individual lists for each type.

You could avoid calling `getAllCollectionWorkflows` and instead get each individual type and add it to to `collectionEntries` here. This way, you also have the lists of each individual type to calculate the count later.

**Review by coverbeck — COMMENTED** [review]
Was here, generally looks good, but going to wait for changes to Kathy's comments before a final review.

File: dockstore-integration-testing/src/test/java/io/dockstore/client/cli/OrganizationIT.java
Comment (denis-yuen) [inline dockstore-integration-testing/src/test/java/io/dockstore/client/cli/OrganizationIT.java:3051]: Remove comment or change to test

File: dockstore-integration-testing/src/test/java/io/dockstore/webservice/ServiceIT.java
Comment (denis-yuen) [inline dockstore-integration-testing/src/test/java/io/dockstore/webservice/ServiceIT.java:139]: Don't need to change this, but should write these tests in a class only for OpenAPI, we have a couple examples. Reduces these package qualifications

**Review by denis-yuen — APPROVED** [review]

File: dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/EntryDAO.java
Comment (coverbeck) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/EntryDAO.java:217]: I suggest either a JavaDoc clarifying this is returning workflows in the JPA sense (because it returns App Tools as well, which is not clear from the method name). Or renaming the method and possibly the JPA query.

I can't think of a good name though, and I'm even having a hard time formulating my comment. :)

File: dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/EntryDAO.java
Comment (coverbeck) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/EntryDAO.java:256]: Same as my other comment.

**Review by coverbeck — APPROVED** [review]
Thanks, this is much improved with the new method names clarifying exactly what is doing what (although I did add one additional suggestion).

File: dockstore-webservice/src/main/resources/swagger.yaml
Comment (hyunnaye) [reply dockstore-webservice/src/main/resources/swagger.yaml:6938]: Ticket created: https://ucsc-cgl.atlassian.net/browse/SEAB-5572

File: dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/EntryDAO.java
Comment (hyunnaye) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/EntryDAO.java:224]: javadoc added

File: dockstore-webservice/src/main/resources/swagger.yaml
Comment (kathy-t) [inline dockstore-webservice/src/main/resources/swagger.yaml:6938]: Since swagger doesn't get updated anymore, I think we should remove this addition to the swagger.yaml. If I understand correctly, this change occurred before you merged the develop branch to this branch

**Review by kathy-t — APPROVED** [review]

**Review by svonworl — APPROVED** [review]

**Pull request merged**
